### ISSUE TITLE
Question about the synchronazation in low-precision kernel

### ISSUE THREAD

**Issue by cat538** [efeslab/Atom#20]
First of all, thank you for your wonderful work! I have a few small questions about the low precision matrix multiplication code.

In `kernels/include/GEMM/Dense_layer_gemm_i4_o16.cuh`, in the kernel `compute_gemm_imma`. I noticed that in the entire kernel, when performing cp.async synchronization, only the statement
```
asm volatile("cp.async.wait_group %0;\n" ::"n"(STAGE - 2));
```
is involved.

My question is, has the correctness of this kernel been checked? Because in my understanding, a complete kernel should involve similar 

`asm volatile("cp.async.wait_group %0;\n" ::"n"(STAGE - 3));`

and `asm volatile("cp.async.wait_group %0;\n" ::"n"(0));`

In other words, we should see statements like "wait for the last two memory transactions to complete" somewhere in the code, right? Of course, it is possible that I have missed some of the code. If you can provide some help in explaining the code, I will be very grateful!

**Comment by happierpig** [comment]
Hi @cat538 ,

Thanks for providing a detailed description of your question.

Sure, our implementation was paired and tested with unit tests of CPU-based calculation. The major misunderstanding here is that each iteration in the main K loop of the GEMM kernel will call:

`asm volatile("cp.async.commit_group;\n" ::);`

which will create a new cp.async-group no matter whether there are any uncommitted cp.async instructions. Please refer to official CUDA docs [here](https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cp-async-commit-group). Therefore, the sync instructions `asm volatile("cp.async.wait_group %0;\n" ::"n"(STAGE - 2));` will guarantee the correct load. Besides, there exists similar open-sourced implementations, check [here](https://github.com/KnowingNothing/MatmulTutorial/blob/main/examples/matmul/this-sm80/matmul-v15.cu).

Hope this can answer your question.

**Comment by cat538** [comment]
Oh this is a really great tip! This makes the code look much cleaner and neater, thanks for the detailed explanation!
